### PR TITLE
PERF: HDFStore __unicode__ method

### LINKED ISSUE #16503: HDFStore __unicode__ very slow for many keys
#### Code Sample, a copy-pastable example if possible
```python

import pandas as pd
store = pd.HDFStore('test.h5', 'w')
for i in range(5000):
    store.put('table_{}'.format(i), pd.DataFrame([i]))

%time str(store)
CPU times: user 26.1 s, sys: 156 ms, total: 26.2 s
Wall time: 26.2 s
```
#### Problem description

The __unicode__ method of HDFStore iterates over all the keys in the file to create the string representation. For larger files this operation becomes extremely slow and dumps an excessive amount of output to the console. 

Worse, this completely bogs down PyCharm's debugger because it calls str(store) for every store that's in scope, on every step.

#### Expected Output

__unicode__ should be a fast operation - just showing the file path would be sufficient. Detailed info on all the keys could be a separate method if needed at all.

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-78-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_CA.UTF-8
LOCALE: en_CA.UTF-8

pandas: 0.19.2
nose: None
pip: 9.0.1
setuptools: 27.2.0
Cython: None
numpy: 1.12.1
scipy: 0.19.0
statsmodels: 0.8.0
xarray: None
IPython: 5.3.0
sphinx: 1.5.3
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2016.10
blosc: None
bottleneck: None
tables: 3.3.0
numexpr: 2.6.2
matplotlib: 2.0.2
openpyxl: None
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.999
httplib2: None
apiclient: None
sqlalchemy: 1.1.9
pymysql: None
psycopg2: 2.7.1 (dt dec pq3 ext lo64)
jinja2: 2.9.6
boto: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by Kiv**
HDFStore __unicode__ now only returns file path info. New info() method has the previous behavior of __unicode__.

```python
import pandas as pd
store = pd.HDFStore('test.h5', 'w')
for i in range(5000):
    store.put('table_{}'.format(i), pd.DataFrame([i]))

# Before
%time str(store)
CPU times: user 26.1 s, sys: 156 ms, total: 26.2 s
Wall time: 26.2 s

# After
%time str(store)
CPU times: user 0 ns, sys: 0 ns, total: 0 ns
Wall time: 40.1 µs
```

 - [x] closes #16503
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

File: pandas/io/pytables.py [hunk 1]
```diff
@@ -1161,6 +1136,35 @@ def copy(self, file, mode='w', propindexes=True, keys=None, complib=None,
 
         return new_store
 
+    def info(self):
+        """return detailed information on the store"""
```
Comment (jreback) [inline pandas/io/pytables.py]: add a versionadded tag.

File: doc/source/whatsnew/v0.21.0.txt [hunk 2]
```diff
@@ -75,7 +75,7 @@ Removal of prior version deprecations/changes
 
 Performance Improvements
 ~~~~~~~~~~~~~~~~~~~~~~~~
-
+- :class:`pandas.HDFStore`'s string representation is now faster and less detailed. For the previous behavior, use ``pandas.HDFStore.info()``. (:issue:`16503`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: move to API breaking changes

File: pandas/io/pytables.py [hunk 3]
```diff
@@ -1161,6 +1136,35 @@ def copy(self, file, mode='w', propindexes=True, keys=None, complib=None,
...
+        output = '%s\nFile path: %s\n' % (type(self), pprint_thing(self._path))
+        if self.is_open:
```
Comment (jreback) [inline pandas/io/pytables.py]: add this to api.rst

**Review by jreback — APPROVED** [review]
lgtm, with some minor comment.

**Comment by jreback** [comment]
ping when fixed / green.

**Comment by gfyoung** [comment]
1) Got a small linting error in `pytables.py` (hence the Travis failure)

2) Can you provide some timing info in the PR given that this is a performance change?

**Comment by Kiv** [comment]
@jreback I think this is ready to go?

File: pandas/tests/io/test_pytables.py [hunk 4]
```diff
@@ -4371,11 +4373,11 @@ def test_multiple_open_close(self):
 
             # single
             store = HDFStore(path)
-            assert 'CLOSED' not in str(store)
+            assert 'CLOSED' not in store.info()
```
Comment (jreback) [inline pandas/tests/io/test_pytables.py]: so you should also test str(store) a bit (u changed all of these to .info())

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/io/test_pytables.py (hunk 4 above)
Comment (Kiv) [inline pandas/tests/io/test_pytables.py]: I added a test that str(store) prints the filename, is that what you mean? There's not much behavior in there anymore to test...

Comment (jreback) [inline pandas/tests/io/test_pytables.py]: point to it it's not clear where it is

Comment (Kiv) [inline pandas/tests/io/test_pytables.py]: https://github.com/pandas-dev/pandas/pull/16514/commits/6faa5a66c005343c71b1ce623bb3ff89affd72b1#diff-5066b062c892df33c29b37a8c1c75f75R422

**Comment by gfyoung** [comment]
@Kiv Some timing info in the PR would be useful for future reference.  Plus a benchmark test would be good here.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@Kiv lgtm. if you would add an asv would be great (you can use your original example, just not as long, maybe 20 nodes or something would be fine).

**Comment by jreback** [comment]
can you add an asv?

**Comment by Kiv** [comment]
My git branch got out of date and I don't know how to update it properly. New version with ASV bench is here:

https://github.com/pandas-dev/pandas/pull/16666

**Comment by jreback** [comment]
in the future see: http://pandas.pydata.org/pandas-docs/stable/contributing.html#combining-commits

you just needed to do:

```
git rebase -i upstream/master
git push yourremove yourbranch -f
```

**Pull request closed without merging**